packages/base/src/commonMain/kotlin/elide/struct/RedBlackTreeMap.kt

```diff
@@ -0,0 +1,141 @@
+
+    override fun add(element: K): Boolean = error("Adding elements to a key set is not supported")
+    override fun addAll(elements: Collection<K>): Boolean = error("Adding elements to a key set is not supported")
+
+    override fun remove(element: K): Boolean = removeNodeByKey(element) != null
+    override fun removeAll(elements: Collection<K>): Boolean {
+      return elements.fold(false) { modified, e -> remove(e) || modified }
+    }
+
+    override fun retainAll(elements: Collection<K>): Boolean {
+      var modified = false
+      for (node in nodes()) if (node.key !in elements) {
+        removeNode(node)
+        modified = true
+      }
+      return modified
+    }
+
+    override fun clear() = reset()
+  }
+
+  /**
+   * A specialized [MutableSet] sharing the tree of its parent map, to be used as a read/write view of its entries.
+   * Operations performed on the set will be reflected on the map immediately at no additional cost.
+   *
+   * Adding entries is not allowed, similar to how the key sets on the JVM's maps forbid additions. As a restriction,
+   * the [iterator] does not support removal.
+   */
+  private inner class EntrySet : MutableSet<MutableEntry<K, V>> {
+    override val size: Int get() = nodeCount
+    override fun isEmpty(): Boolean = nodeCount == 0
+
+    override fun iterator(): MutableIterator<MutableEntry<K, V>> {
+      val inner = nodes().iterator()
+      return object : MutableIterator<MutableEntry<K, V>> {
+        override fun hasNext(): Boolean = inner.hasNext()
+        override fun next(): MutableEntry<K, V> = inner.next()
+        override fun remove() = error("Removing values while traversing the set is not supported")
+      }
+    }
+
+    override fun contains(element: MutableEntry<K, V>): Boolean = findNodeByKey(element.key) != null
+    override fun containsAll(elements: Collection<MutableEntry<K, V>>): Boolean = elements.all(::contains)
+
+    override fun add(element: MutableEntry<K, V>): Boolean {
+      error("Adding elements to an entry set is not supported")
+    }
+
+    override fun addAll(elements: Collection<MutableEntry<K, V>>): Boolean {
+      error("Adding elements to an entry set is not supported")
+    }
+
+    override fun remove(element: MutableEntry<K, V>): Boolean = removeNodeByKey(element.key) != null
+    override fun removeAll(elements: Collection<MutableEntry<K, V>>): Boolean {
+      return elements.fold(false) { modified, e -> remove(e) || modified }
+    }
+
+    override fun retainAll(elements: Collection<MutableEntry<K, V>>): Boolean {
+      var modified = false
+      for (node in nodes()) if (node !in elements) {
+        removeNode(node)
+        modified = true
+      }
+      return modified
+    }
+
+    override fun clear() = reset()
+  }
+
+  override val keys: MutableSet<K> by lazy(::KeySet)
+  override val entries: MutableSet<MutableEntry<K, V>> by lazy(::EntrySet)
```

Comment:
Sets returned by `keys` and `entries` share the same nodes as the map itself, no additional collections are allocated, and removing keys/entries from the sets will also update the map (following JVM maps behavior)